provider/datagen/src/driver.rs

```diff
@@ -11,14 +11,230 @@ use icu_locid_transform::LocaleFallbacker;
+    /// Stylized on the CLI as: "@en-US"
+    pub fn single(langid: LanguageIdentifier) -> Self {
+        Self {
+            langid: Some(langid),
+            include_ancestors: false,
+            include_descendants: false,
+        }
+    }
+
+    /// The family containing all locales.
+    ///
+    /// Stylized on the CLI as: "full"
+    pub fn full() -> Self {
+        Self {
+            langid: None,
+            include_ancestors: false,
+            include_descendants: true,
+        }
+    }
+}
+
+impl fmt::Display for LocaleFamily {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match (
+            &self.langid,
+            self.include_ancestors,
+            self.include_descendants,
+        ) {
+            (Some(langid), true, true) => write!(f, "{}", langid),
+            (Some(langid), true, false) => write!(f, "^{}", langid),
+            (Some(langid), false, false) => write!(f, "@{}", langid),
+            (Some(_), false, true) => unreachable!(),
+            (None, _, _) => write!(f, "full"),
+        }
+    }
+}
+
+/// Options bag configuring locale inclusion and behavior when runtime fallback is enabled.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Default)]
+#[non_exhaustive]
+pub struct FallbackOptions {
+    /// The location in code where fallback will be performed at runtime.
+    ///
+    /// If not set, determined by the exporter: internal fallback is used if the exporter
+    /// supports internal fallback.
+    pub runtime_fallback_location: Option<RuntimeFallbackLocation>,
+    /// The aggressiveness of deduplication of data payloads.
+    ///
+    /// If not set, determined by `runtime_fallback_location`. If internal fallback is enabled,
+    /// a more aggressive deduplication strategy is used.
+    pub deduplication_strategy: Option<DeduplicationStrategy>,
+}
+
+#[derive(Debug, Clone)]
+enum LocalesWithOrWithoutFallback {
+    WithFallback {
+        locales: HashSet<LocaleFamily>,
+        options: FallbackOptions,
+    },
+    WithoutFallback {
+        locales: HashSet<LanguageIdentifier>,
+    },
+}
+
+impl fmt::Display for LocalesWithOrWithoutFallback {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::WithFallback { locales, options } => {
+                let mut sorted_locales =
+                    locales.iter().map(ToString::to_string).collect::<Vec<_>>();
+                sorted_locales.sort();
+                let start = match options.runtime_fallback_location {
+                    None => unreachable!(),
+                    Some(RuntimeFallbackLocation::Internal) => "with internal fallback",
+                    Some(RuntimeFallbackLocation::External) => "with external fallback",
+                };
+                write!(f, "{start}, ")?;
+                match options.deduplication_strategy {
+                    Some(x) => x.describe(f)?,
+                    None => unreachable!(),
+                }
```

Comment:
nit: consider
```suggestion
                options.deduplication_strategy.expect("resolved").describe(f)?;
```

Reply:
Rewrote this to be after the option resolution and remove the unreachable case.